zeroex/orderwatch/order_watcher.go

```diff
@@ -99,50 +93,382 @@ func New(meshDB *meshdb.MeshDB, blockWatcher *blockwatch.Watcher, orderValidator
+	}()
+	mainLoopErrChan := make(chan error, 1)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		mainLoopErrChan <- w.mainLoop(innerCtx)
+	}()
+	cleanupLoopErrChan := make(chan error, 1)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		cleanupLoopErrChan <- w.cleanupLoop(innerCtx)
+	}()
+
+	// If any error channel returns a non-nil error, we cancel the inner context
+	// and return the error. Note that this means we only return the first error
+	// that occurs.
+	select {
+	case err := <-expirationErrChan:
+		if err != nil {
+			cancel()
+			return err
+		}
+	case err := <-mainLoopErrChan:
+		if err != nil {
+			cancel()
+			return err
+		}
+	case err := <-cleanupLoopErrChan:
+		if err != nil {
+			cancel()
+			return err
+		}
+	}
 
-	w.isSetup = true
+	// Wait for all goroutines to exit. If we reached here it means we are done
+	// and there are no errors.
+	wg.Wait()
 	return nil
 }
 
-// Stop closes the block subscription, stops the event, expiration watcher and the cleanup worker.
-func (w *Watcher) Stop() error {
-	w.setupMux.Lock()
-	if !w.isSetup {
-		w.setupMux.Unlock()
-		return errors.New("Cannot teardown before calling Setup()")
+func (w *Watcher) mainLoop(ctx context.Context) error {
+	// Set up the channel used for subscribing to block events.
+	blockEvents := make(chan []*blockwatch.Event, 10)
+	w.blockSubscription = w.blockWatcher.Subscribe(blockEvents)
+
+	for {
+		select {
+		case <-ctx.Done():
+			w.blockSubscription.Unsubscribe()
+			close(blockEvents)
+			return nil
+		case expiredOrders := <-w.expirationWatcher.ExpiredItems():
+			w.handleExpiration(expiredOrders)
+		case err := <-w.blockSubscription.Err():
+			logger.WithFields(logger.Fields{
+				"error": err.Error(),
+			}).Error("block subscription error encountered")
+		case events := <-blockEvents:
+			if err := w.handleBlockEvents(events); err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func (w *Watcher) cleanupLoop(ctx context.Context) error {
+	start := time.Now()
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+
+		// Wait minCleanupInterval before calling cleanup again. Since
+		// we only start sleeping _after_ cleanup completes, we will never
+		// have multiple calls to cleanup running in parallel
+		time.Sleep(minCleanupInterval - time.Since(start))
+		start = time.Now()
+		if err := w.cleanup(ctx); err != nil {
+			return err
+		}
 	}
-	w.setupMux.Unlock()
+}
 
-	// Stop event subscription
-	w.blockSubscription.Unsubscribe()
+func (w *Watcher) handleExpiration(expiredOrders []expirationwatch.ExpiredItem) {
+	orderEvents := []*zeroex.OrderEvent{}
+	for _, expiredOrder := range expiredOrders {
+		order := &meshdb.Order{}
+		err := w.meshDB.Orders.FindByID(common.HexToHash(expiredOrder.ID).Bytes(), order)
+		if err != nil {
+			logger.WithFields(logger.Fields{
+				"error":     err.Error(),
+				"orderHash": expiredOrder.ID,
+			}).Warning("Order expired that was no longer in DB")
```

Comment:
Should we keep this as a `Warning` or lower this to a `Trace`? It is expected to happen from time to time.

Reply:
Sure. Good idea.